### PR TITLE
Improve Fixture support for Active Storage

### PR DESCRIPTION
**Pull request by seanpdoyle**
Inspired by [76b33aa][], this commit extends the Active Storage
documentation to elaborate on how to declare fixtures.

In support of that, also introduce the `ActiveStorage::FixtureSet.blob`
method for injecting in-line [ActiveStorage::Blob][] attributes directly
into fixture YAML.

[76b33aa]: https://github.com/rails/rails/commit/76b33aa3d102ad4ff3f766ac2d3a711d73aaee82
[ActiveStorage::Blob]: https://edgeapi.rubyonrails.org/classes/ActiveStorage/Blob.html

### CODE REVIEW

**Comment by sedubois** [comment]
@seanpdoyle this is great. So far I've been making fixtures this way:

```yml
# test/fixtures/active_storage/blobs.yml

cover_image_blob:
  service_name: test
  key: euk5bxrw3jfay143dv9a5crb02ln
  filename: cover.jpg
  content_type: image/jpg
  byte_size: <%= File.size('tmp/storage/eu/k5/euk5bxrw3jfay143dv9a5crb02ln') %>
  checksum: <%= Digest::MD5.file('tmp/storage/eu/k5/euk5bxrw3jfay143dv9a5crb02ln').base64digest %>

cover_image_variant_150x150:
  <<: *DEFAULTS
  key: 8qp71n7snpep9i7s4m72otrnreeg
  filename: cover.jpg
  content_type: image/png
  byte_size: <%= File.size('tmp/storage/8q/p7/8qp71n7snpep9i7s4m72otrnreeg') %>
  checksum: <%= Digest::MD5.file('tmp/storage/8q/p7/8qp71n7snpep9i7s4m72otrnreeg').base64digest %>
  metadata: '{"identified":true,"width":150,"height":84,"analyzed":true}'

# test/fixtures/active_storage/blobs.yml

blog_post_cover_image:
  name: cover_image
  record: blog_post (Post)
  blob: cover_image_blob

# test/fixtures/active_storage/variant_records.yml

cover_image_150x150:
  blob: cover_image_variant_150x150
  variation_digest:
    <%= ActiveStorage::Variation.new(resize_to_limit: [150, 150]).digest %>
```

(Even if VariantRecords are generated automatically, I added a fixture for these too to ensure I was not breaking the database during some migrations.)

File: activestorage/lib/active_storage/fixture_set.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    # the file to the Service
+    #
+    # === Examples
+    #
+    #   # tests/fixtures/action_text/blobs.yml
+    #   second_thumbnail_blob: <%= ActiveStorage::FixtureSet.blob(
+    #    filename: "second.svg",
+    #    content_type: "image/svg+xml",
+    #   ) %>
+    def self.blob(filename:, **attributes)
```
Comment (seanpdoyle) [inline activestorage/lib/active_storage/fixture_set.rb:52]: @georgeclaghorn this branch is a corollary to https://github.com/rails/rails/pull/41062.

**Comment by sedubois** [comment]
With this PR, does the 'test' storage service in storage.yml becomes useless? Should it be removed?

**Comment by seanpdoyle** [comment]
> With this PR, does the 'test' storage service in storage.yml becomes useless? Should it be removed?

@sedubois I'm not sure I follow. What makes it useless?

**Comment by sedubois** [comment]
Oh sorry I misunderstood, the files are initially in test/fixtures but copied to tmp/storage when running tests.

File: activestorage/lib/active_storage/fixture_set.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  # \Fixtures are a way of organizing data that you want to test against; in
+  # short, sample data.
+  #
+  # To learn more about fixtures, read the ActiveRecord::FixtureSet
+  # documentation.
+  #
+  # === YAML
+  #
+  # Like other Active Record-backed models, ActiveStorage::Attachment and
+  # ActiveStorage::Blob records inherit # from ActiveRecord::Base instances and
```
Comment (sedubois) [inline activestorage/lib/active_storage/fixture_set.rb:13]:
```suggestion
  # ActiveStorage::Blob records inherit from ActiveRecord::Base instances and
```

**Comment by sedubois** [comment]
I tried to use `ActiveStorage::FixtureSet` but encounter some errors when inspecting the blob (`post.cover_image.blob`/`post.cover_image.attachment.blob` throw `JSON::ParserError`). Because of this I also can't call `analyze` like I used to do in my tests and thus I can't access the blob's metadata. @seanpdoyle would you know what I might be doing wrong?

```
# test/fixtures/files/cover.jpg: contains valid JPEG image

# lib/active_storage/fixture_set.rb: copied from this PR
```

```yml
# test/fixtures/active_storage/blobs.yml
cover_image: <%= ActiveStorage::FixtureSet.blob filename: "cover.jpg" %>

# test/fixtures/active_storage/attachments.yml
post_1_cover_image:
  name: cover_image
  record: post_1 (Post)
  blob: cover_image

# test/fixtures/posts.yml
post_1: {}
```

```rb
# app/models/post.rb
class Post < ApplicationRecord
  has_one_attached :cover_image
end

# test/test_helper.rb
require "active_storage/fixture_set"
ActiveStorage::FixtureSet.file_fixture_path = File.expand_path("fixtures/files", __dir__)
# ...

# test/models/post_test.rb
require 'test_helper'

class PostTest < ActiveSupport::TestCase
  test "cover_image is attached" do
    post = posts(:blog_post_1)

    assert post.cover_image.attached?

    puts post.cover_image.attachment.inspect
    # #<ActiveStorage::Attachment id: "1218d74d-a0be-503b-882d-ada51555fd34", name: "cover_image", record_type: "Post", created_at: "2021-01-19 07:10:52.281813000 +0000", record_id: "d22e4e23-91f8-5b2a-9721-5ec1442ee985", blob_id: "d0032c61-4923-5b44-a662-42c33d4d01d3">

    puts post.cover_image.attachment.blob.inspect # <<- throws error:
    # Minitest::UnexpectedError: JSON::ParserError: 809: unexpected token at '---
    # identified: true
    # '

    post.cover_image.attachment.analyze # <<- also throws same error
  end
end
```

**Comment by seanpdoyle** [comment]
Thanks for reporting that @sedubois! I think I've resolved it.

**Comment by sedubois** [comment]
My pleasure @seanpdoyle. NB: there is still a [broken test](https://buildkite.com/rails/rails/builds/74328#447cf728-58a3-4413-ae31-a8c405f6597b) (ActiveStorage::AttachmentTest#test_signed_blob_ID_backwards_compatibility), don't know why...

**Comment by dhh** [comment]
Actually, yes, I prematurely merged this before that test was fixed. @seanpdoyle can you sort that one?

**Comment by sedubois** [comment]
@seanpdoyle would you also mind showing how to specify the service in which to store the blob?

**Comment by sedubois** [comment]
Might also be worth documenting how to configure the service:

```rb
second_thumbnail_blob: <%= ActiveStorage::FixtureSet.blob(
  filename: "second.svg",
  content_type: "image/svg+xml",
  service_name: "public"
)
```

**Pull request merged**